compiler-core/src/language_server/code_action.rs

```diff
@@ -7562,3 +7562,374 @@ impl<'ast> ast::visit::Visit<'ast> for RemovePrivateOpaque<'ast> {
+            .src_span_to_lsp_range(outer_clause_span)
+            .start
+            .character;
+        let indentation = " ".repeat(pattern_nesting as usize);
+
+        self.edits.replace(
+            outer_clause_span,
+            new_clauses.join(&format!("\n{indentation}")),
+        );
+
+        let mut action = Vec::with_capacity(1);
+        CodeActionBuilder::new("Collapse nested case")
+            .kind(CodeActionKind::REFACTOR_REWRITE)
+            .changes(self.params.text_document.uri.clone(), self.edits.edits)
+            .preferred(false)
+            .push_to(&mut action);
+        action
+    }
+
+    fn code_at(&self, span: SrcSpan) -> &str {
+        self.module
+            .code
+            .get(span.start as usize..span.end as usize)
+            .expect("location must be valid")
+    }
+}
+
+impl<'ast> ast::visit::Visit<'ast> for CollapseNestedCase<'ast> {
+    fn visit_typed_clause(&mut self, clause: &'ast ast::TypedClause) {
+        let ast::TypedClause {
+            pattern,
+            alternative_patterns,
+            then,
+            location,
+            guard,
+        } = clause;
+
+        // We can only flatten a case expression in a very specific case:
+        // - This pattern may be introducing multiple variables,
+        // - The expression following this branch must be a case, and
+        // - It must be matching on one of those variables
+        //
+        // For example:
+        //
+        // ```gleam
+        // Wibble(a, b, 1) -> case a { ... }
+        // Wibble(a, b, 1) -> case b { ... }
+        // ```
+
+        // We can't have any alternative patterns...
+        if alternative_patterns.is_empty()
+            // ...and the then clause must be a single case expression...
+            && let Some(TypedExpr::Case {
+                subjects, clauses, ..
+            }) = then.expression_flattening_block()
+            // ...that is matching a single variable...
+            && let [TypedExpr::Var { name, .. }] = subjects.as_slice()
+            // ...and that variable must be one the variables we brought into
+            // scope in this branch.
+            && let Some(variable) = pattern
+                .iter()
+                .flat_map(|pattern| pattern.bound_variables())
+                .find(|variable| variable.name() == *name)
+        {
+            // There's one last condition to trigger the code action: we must
+            // actually be with the cursor over the pattern or the nested case
+            // expression!
+            //
+            // ```gleam
+            // case wibble {
+            //   Ok(a) -> case a {
+            // //^^^^^^^^^^^^^^^ Anywhere over here!
+            //   }
+            // }
+            // ```
+            //
+            let pattern_location = pattern
+                .first()
+                .map_or(*location, |first| first.location())
+                .merge(&pattern.last().map_or(*location, |end| end.location()));
```

Comment:
Sorry, might be being slow here. When is it that there might not be any patterns?

Reply:
I was being cautious, since first return an option. But I could just `expect` it!